src/main/java/uk/ac/ebi/subs/processing/dispatcher/DispatcherServiceImpl.java

```diff
@@ -177,6 +188,33 @@ public void insertReferencedSamples(SubmissionEnvelope submissionEnvelope) {
         submissionEnvelope.getSupportingSamples().addAll((Set<? extends Sample>) refLookupService.lookupRefs(assaySampleRefs));
     }
 
+    @Override
+    public void insertUploadedFiles(SubmissionEnvelope submissionEnvelope) {
+        String submissionId = submissionEnvelope.getSubmission().getId();
+
+        List<UploadedFile> uploadedFiles = submissionEnvelope.getAssayData().stream().map(
+                AssayData::getFiles
+        ).collect(Collectors.toList())
+            .stream().flatMap(List::stream).collect(Collectors.toList())
+                .stream().map(
+                    fileRef -> {
+                        String fileName = fileRef.getName();
+                        File file = fileRepository.findByFilenameAndSubmissionId(fileName, submissionId);
```

Comment:
This has the potential to be a lot of database lookups - could we get all the uploaded files in one go?

Reply:
Thanks for this catch and perfectly agree. Bad code 👎

Reply:
Rewrote